File: 백준/Silver/24060. 알고리즘 수업 － 병합 정렬 1/알고리즘 수업 － 병합 정렬 1.py
def merge(nums, s, q, e):
    global cnt, num
    i = s
    j = q+1
    t = 0
    tmp = [0] * (e-s+1)
    while (i <= q and j <= e):
        if (nums[i] <= nums[j]):
            tmp[t] = nums[i]
            t += 1
            i += 1
        else:
            tmp[t] = nums[j]
            t += 1
            j += 1

    while (i <= q):
        tmp[t] = nums[i]
        t += 1
        i += 1

    while (j <= e):
        tmp[t] = nums[j]
        t += 1
        j += 1

    i = s
    t = 0
    while (i <= e):
        nums[i] = tmp[t]
        cnt += 1
        if cnt == K:
            num = tmp[t]
        t += 1
        i += 1



def merge_sort(arr, s, e):
    if (s < e):
        q = (s + e) // 2
        merge_sort(arr, s, q)
        merge_sort(arr, q+1, e)
        merge(arr, s, q, e)


N, K = map(int, input().split())
nums = list(map(int, input().split()))
cnt = 0
num = -1

merge_sort(nums, 0, N-1)

if cnt < K:
    print(-1)
else:
    print(num)

File: 백준/Silver/24060. 알고리즘 수업 － 병합 정렬 1/알고리즘 수업 － 병합 정렬 1.java
import java.io.*;
import java.util.StringTokenizer;
public class Main {
    static int N, K;
    static int[] nums;
    static int result;
    static int cnt;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());
        nums = new int[N];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            nums[i] = Integer.parseInt(st.nextToken());
        }
        result = -1;
        merge_sort(0, N-1);
        System.out.println(result);
    }
    static public void merge_sort(int p, int r) {
        if (cnt > K) {
            return;
        }

        if (p < r) {
            int q = (p + r) / 2;
            merge_sort(p, q);
            merge_sort(q+1, r);
            merge(p, q, r);
        }
    }
    static public void merge(int p, int q, int r) {
        int i = p;
        int j = q + 1;
        int t = 0;
        int[] tmp = new int[r-p+1];
        while (i <= q & j <= r) {
            if (nums[i] <= nums[j]) {
                tmp[t++] = nums[i++];


            } else {
                tmp[t++] = nums[j++];
            }
        }

        while (i <= q) {
            tmp[t++] = nums[i++];
        }
        while (j <= r) {
            tmp[t++] = nums[j++];
        }
        i = p;
        t = 0;
        while (i <= r) {
            nums[i] = tmp[t];
            cnt++;
            if (cnt == K) {
                result = tmp[t];
                break;
            }
            t++;
            i++;
        }
    }
}
